### PR TITLE
Support Spark YARN cluster deployment mode

### PR DESCRIPTION
**Pull request by xjzrc**
Support Spark YARN cluster deployment mode

### CODE REVIEW

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 1]
```diff
@@ -65,6 +66,14 @@ class KyuubiSessionImpl(
   private var client: TCLIService.Client = _
   private var remoteSessionHandle: TSessionHandle = _
 
+  private val isClusterMode = sessionConf.getOption(SPARK_SUBMIT_DEPLOY_MODE)
+    .map(_.toLowerCase(Locale.ROOT)).contains("cluster")
+
+  if (isClusterMode) {
+    sessionConf.set(SPARK_YARN_MAX_APP_ATTEMPTS, "1")
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:73]: Why should it be 1?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 2]
```diff
@@ -65,6 +66,14 @@ class KyuubiSessionImpl(
...
+    sessionConf.set(SPARK_YARN_SUBMIT_WAIT_APP_COMPLETION, "false")
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: Is it necessary?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 3]
```diff
@@ -95,7 +104,7 @@ class KyuubiSessionImpl(
         var sh = getServerHost
         val started = System.currentTimeMillis()
         while (sh.isEmpty) {
-          if (process.waitFor(1, TimeUnit.SECONDS)) {
+          if (!isClusterMode && process.waitFor(1, TimeUnit.SECONDS)) {
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:107]: does this mean that cluster mode never fails with client exceptions/errors?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 1 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:73]: Prevent AM restarts when the engine reaches its maximum idle time and terminated.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 2 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: In cluster mode, clients should not be resident.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 3 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:107]: yes, the engine will start in AM.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 1 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:73]: AM will restart when it terminated w/ FINISHED state?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 3 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:107]: You did not get my point

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 2 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: what is wrong if we keep the client alive?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 1 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:73]: enen

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 2 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: No, but it consumes kyuubi service's resources.

Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: so i think set `spark.yarn.submit.waitAppCompletion=false` is necessary.

Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: Let us keep it configurable in case of any regression happen?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 3 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:107]: In client mode, we can use `process.waitFor` to return true to determine engine startup exceptions. But in cluser mode, if   'spark.yarn.submit.waitAppCompletion=false' is set , `process.waitFor` always return true.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 1 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:73]: What makes it different from other regular spark applications? I guess they are the same

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 3 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:107]: The client exception is useful. for example, your YARN cluster is crashed, or maybe the server side is configured wrong or the user overwrites some setting, then Kyuubi failed to start the engine proc, we need to capture the err.

**Comment by yaooqinn** [comment]
we need some end to end case to test this feature

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 1 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:73]: I made a mistake,  AM didn't w/ FINISHED state when it terminated. I will fix it.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 2 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:74]: ok

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 3 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala]: @yaooqinn I've fixed it.

**Comment by xjzrc** [comment]
> we need some end to end case to test this feature?
I packaged this for testing in our YARN cluster. How do we do unit testing locally.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 4]
```diff
@@ -66,6 +66,13 @@ case class KyuubiConf(loadSysDefault: Boolean = true) extends Logging {
     this
   }
 
+  def setIfMissing(key: String, value: String): KyuubiConf = {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:69]: unnecessary change？

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 5]
```diff
@@ -96,9 +96,13 @@ class KyuubiSessionImpl(
         info(s"Launching SQL engine: $builder")
         var sh = getServerHost
         val started = System.currentTimeMillis()
+        var exitValue: Option[Int] = None
```
Comment (yaooqinn) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:99]: unnecessary definition？

**Comment by yaooqinn** [comment]
> > we need some end to end case to test this feature?
> 
> I packaged this for testing in our YARN cluster. How do we do unit testing locally.

That is OK, but I didn't know that before you told me.

BTW, if you verified locally in your yarn cluster, could you please provide the screenshot or anything else that can help me confirm that.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 4 above)
Comment (xjzrc) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:69]: sorry, I forgot to remove it.

File: kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 5 above)
Comment (xjzrc) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:99]: necessary.  there is no need to execute `process.waitFor`, when the process exits normally, eg: cluster mode and 'spark.yarn.submit.waitAppCompletion=false'

Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:99]: If the process exits already, it won't wait.

**Comment by xjzrc** [comment]
> > > we need some end to end case to test this feature?
> > 
> > 
> > I packaged this for testing in our YARN cluster. How do we do unit testing locally.
> 
> That is OK, but I didn't know that before you told me.
> 
> BTW, if you verified locally in your yarn cluster, could you please provide the screenshot or anything else that can help me confirm that.
kyuubi server log:
image
client log:
image
AM log:
image
image
